src/coffeeShops/useCoffeeShopsFetch.ts

```diff
@@ -0,0 +1,84 @@
+import { useEffect, useState } from "react";
+import { coffeeShopsUrl, tokenUrl } from "../api";
+import fetchData, { FetchResult } from "../fetchData";
+import { CoffeeShop, Token } from "./models";
+
+const retryCount = 3;
+const retryInterval = 1000;
+const fetchTimeout = 2000;
+
+type CoffeeShopsFetchResult = { token?: string } & FetchResult<CoffeeShop[]>;
+
+export default function useCoffeeShopsFetch() {
+  const [fetchResult, setFetchResult] = useState<CoffeeShopsFetchResult>({});
+  const [isLoading, setIsLoading] = useState(true);
+
+  useEffect(() => {
+    (async () => {
+      setFetchResult(await fetchCoffeeShopsWithRetry(retryCount));
+      setIsLoading(false);
+    })();
+  }, []);
+
+  return { ...fetchResult, isLoading };
+}
+
+async function fetchCoffeeShopsWithRetry(
+  retries: number,
+  prevToken?: string
+): Promise<CoffeeShopsFetchResult> {
+  const fetchResult = await fetchCoffeeShops(prevToken);
+
+  retries = retries - 1;
+
+  if (
+    fetchResult.httpStatus === 200 ||
```

Comment:
These condition grew a bit too much that it becomes hard to follow the full thing. Would it be nicer to have const with meaningful name?

Reply:
I agree.